rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,315 @@
+# SlateDB Checkpoints and Clones
+
+Status: Under Review
+
+Authors:
+* [Rohan Desai](https://github.com/rodesai)
+
+## Current Implementation
+
+SlateDB runs a Writer, a Compactor - optionally in a separate process, and a Garbage Collector (GC) - also optionally in a separate process. The writer writes data to L0 SSTs. The Compactor then goes and compacts these L0 SSTs into Sorted Runs (SR) under the “compacted” directory, and then further compacts those SRs as they accumulate. Updates to the current set of L0s and Sorted Runs are synchronized between the Writer and Compactor through updates to the centralized manifest. When reading, the Writer gets a reference to a copy of the metadata for the current set of sorted runs and searches them for the requested data. Meanwhile, the GC scans the SSTs in Object Storage and deletes any SSTs that are not referenced by the current manifest and are older than some threshold.
```

Comment:
Nit: GC currently deletes items that are SSTs older than a threshold, there is an issue tracking to delete it `threshold` minutes after it has become inactive. #204 . This aspect is relevant for checkpoints.

Reply:
Ack. I think we can actually build on this proposal to implement the functionality proposed in #204. Left a comment there.